Migrated user guide notebooks to MyST-NB and added sphinx extension

mmccarty opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please mark your pull request as Draft.
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove it from "Draft" and make it "Ready for Review".
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

   If assistance is required to complete the functionality, for example when the
   C/C++ code of a feature is complete but Python bindings are still required,
   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by
   merging the target branch into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

This PR adds [MyST-NB](https://myst-nb.readthedocs.io/en/latest/index.html) to the sphinx docs build. This will allow the notebooks to be executed when the docs are built to ensure they stay up-to-date with the latest version of the software.

- [x] Conda env yml updates
- [x] Sphinx conf.py updates to enable and configure `myst-nb`
- [x] Minor fixes to the `10min.ipynb` notebook to find test data file path.

shwina commented:
@mmccarty there's a conflict here due to https://github.com/rapidsai/cudf/pull/10690 modifying `guide-to-udfs.ipynb`. However this PR deleted that file and replaced it with `guide-to-udfs.md`.

I incorporated those changes into `guide-to-udfs.md` as part of my downstream PR https://github.com/rapidsai/cudf/pull/10691 -- you can just steal it from [here](https://github.com/rapidsai/cudf/blob/1fad8a251ab9a23310718eeb7dc06ce2299170d0/docs/cudf/source/user_guide/guide-to-udfs.md).

mmccarty commented:
> @mmccarty there's a conflict here due to #10690 modifying guide-to-udfs.ipynb. However this PR deleted that file and replaced it with guide-to-udfs.md.

Reverted `.ipynb` to `.md` conversion and rebased with `branch-22.04` to resolve this conflict.

ajschmidt8 reviewed (commented):
@mmccarty, can you add `sphinx-autobuild` to https://github.com/rapidsai/integration/blob/branch-22.06/conda/recipes/rapids-doc-env/meta.yaml and `myst-nb` to https://github.com/rapidsai/integration/blob/branch-22.06/conda/recipes/rapids-notebook-env/meta.yaml?

mmccarty commented:
Will do! Thanks @ajschmidt8

mmccarty commented:
@ajschmidt8 I'm pretty sure those dependencies are only needed in the docs environment. Where we ended up should not impact notebook execution in Jupyter.

shwina commented:
@galipremsagar looks related to https://github.com/executablebooks/jupyter-book/issues/1611 perhaps?

galipremsagar commented:
> @galipremsagar looks related to [executablebooks/jupyter-book#1611](https://github.com/executablebooks/jupyter-book/issues/1611) perhaps?

Yes, to fix it we will have to add the following rule in https://github.com/rapidsai/cudf/blob/branch-22.06/docs/cudf/source/_static/params.css

```css
.output.text_html {
    overflow-x: scroll;
}
```

shwina commented on docs/cudf/source/conf.py:
```diff
@@ -46,10 +46,13 @@
     "IPython.sphinxext.ipython_directive",
-    "nbsphinx",
     "PandasCompat",
+    "myst_nb",
```
Do we also want to add  [`myst_parser` here to potentially parse `.md` files](https://www.sphinx-doc.org/en/master/usage/markdown.html)?

Or I can just do this in a follow-up.

mmccarty replied:
I'm +1 for that, but I think it can probably be in a follow PR that adds MyST `.md` files.

galipremsagar commented on docs/cudf/source/_static/params.css:
```diff
@@ -52,4 +52,8 @@ table.io-supported-types-table thead{
+}
+
+.output.text_html {
+    overflow: auto;
 }
```
```suggestion
    overflow: auto;
}

```

mmccarty commented on conda/environments/cudf_dev_cuda11.5.yml:
```diff
@@ -54,6 +54,10 @@ dependencies:
   - sphinx-copybutton
+  - sphinx-autobuild
+  - myst-nb
+  - scipy
```
Yes, the `10min-cudf-cupy.ipynb` notebook makes a call to `cupy` that requires `scipy`.

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [21], in <cell line: 2>()
      1 sparse_data = cudf_to_cupy_sparse_matrix(df)
----> 2 print(sparse_data)

File ~/miniconda3/envs/cudf_dev/lib/python3.8/site-packages/cupyx/scipy/sparse/base.py:60, in spmatrix.__str__(self)
     57 def __str__(self):
     58     # TODO(unno): Do not use get method which is only available when scipy
     59     # is installed.
---> 60     return str(self.get())

File ~/miniconda3/envs/cudf_dev/lib/python3.8/site-packages/cupyx/scipy/sparse/csc.py:64, in csc_matrix.get(self, stream)
     50 """Returns a copy of the array on host memory.
     51 
     52 .. warning::
   (...)
     61 
     62 """
     63 if not _scipy_available:
---> 64     raise RuntimeError('scipy is not available')
     65 data = self.data.get(stream)
     66 indices = self.indices.get(stream)

RuntimeError: scipy is not available
RuntimeError: scipy is not available
```

mmccarty commented:
rerun tests

shwina commented:
Looking into the CI failures here.

shwina commented:
rerun tests

shwina commented:
rerun tests

vyasr commented on docs/cudf/source/_static/params.css:
```diff
@@ -52,4 +52,8 @@ table.io-supported-types-table thead{
\ No newline at end of file
+}
+
+.output.text_html {
+    overflow: auto;
+}
```
```suggestion
/* Needed to resolve https://github.com/executablebooks/jupyter-book/issues/1611 */ 
.output.text_html {
    overflow: auto;
}
```

vyasr reviewed (approved):
One minor request to add a comment, otherwise LGTM.

vyasr commented:
I committed my proposed change since it looks like this PR is otherwise ready to merge pending tests passing.

shwina commented:
@gpucibot merge
